### PR TITLE
Fix some of the content deopt cases

### PR DESCRIPTION
**Pull request by chadhietala**
This fixes some of the content deopt cases by emitting the operations to effectively do the following:

```
{{if isComponentReferent (component foo.bar) foo.bar}}
```

This also fixes an issue where the deopt instructions were not being reused, meaning we were compiling new instructions for the deopt effectively creating a slow memory leak.

### CODE REVIEW

File: packages/@glimmer/runtime/lib/compiled/opcodes/builder.ts [hunk 1]
```diff
@@ -560,6 +562,52 @@ export default class OpcodeBuilder extends BasicOpcodeBuilder {
...
+    this.return();
+
+    this.label('END');
+
+    this.popFrame();
+
+    this.stopLabels();
+  }
+
   guardedCautiousAppend(expression: WireFormat.Expression) {
```
Comment (chancancode) [inline packages/@glimmer/runtime/lib/compiled/opcodes/builder.ts]: Are these used?

File: packages/@glimmer/runtime/lib/compiled/opcodes/builder.ts [hunk 2]
```diff
@@ -560,6 +562,52 @@ export default class OpcodeBuilder extends BasicOpcodeBuilder {
...
+
+    this.returnTo('END');
+
+    expr(expression, this);
+
+    this.dup();
+
+    this.test((reference) => {
+      return IsComponentDefinitionReference.create(reference);
+    });
```
Comment (chancancode) [inline packages/@glimmer/runtime/lib/compiled/opcodes/builder.ts]: This could just be `this.test(IsComponentDefinitionReference.create)` right?

Comment (chadhietala) [inline packages/@glimmer/runtime/lib/compiled/opcodes/builder.ts]: Yes

File: packages/@glimmer/runtime/lib/compiled/opcodes/builder.ts (hunk 1 above)
Comment (chadhietala) [inline packages/@glimmer/runtime/lib/compiled/opcodes/builder.ts]: I think Kris might have removed these in a PR after this.

**Pull request merged**